cpp/ql/src/semmle/code/cpp/controlflow/ControlFlowGraph.qll

```diff
@@ -88,14 +105,9 @@ private cached module Cached {
   cached
   predicate reachable(ControlFlowNode n)
```

Comment:
Have you considered rewriting this to use a simple implementation of basic blocks? I do something like that for C#, and since we don't care about the order of the elements inside the basic blocks, you can use `fastTC` in the construction of the simple basic blocks (see https://github.com/Semmle/ql/blob/master/csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll#L3989).

Reply:
I have a few branches lying around where `reachable` is computed with basic blocks instead of control-flow nodes. They give significant speed-ups, and I'll make a PR out of one of them some time this quarter.

We can't use `fastTC` here as the recursion is not actually a TC. It's hard to see from the source, but you can see it more clearly in #170. In `reachable(n.getAPredecessor())`, the `getAPredecessor` predicate recursively depends on `reachable`, so this is a join with a recursive call on both sides of the `and`. That's part of why this predicate is hard to optimise. The difficult recursion structure is needed because it's an interprocedural analysis that considers code to be unreachable when it comes after a call to a function whose exit point is unreachable.

Reply:
I see, thanks for the clarification.